### ISSUE TITLE
Deriving Clone on ObjectSubclass is unsound

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gtk-rs#46]
It would create a clone of just the instance private struct. Calling `ObjectSubclass::get_instance()` on it would then get a reference to some random memory.

This needs to be prevented somehow. Maybe via a cleverly placed [`assert_not_impl_all`](https://docs.rs/static_assertions/1.1.0/static_assertions/macro.assert_not_impl_all.html) in the `glib_object_subclass!` macro.

**Comment by YaLTeR** [comment]
Even something like this is UB I guess:
```rust
impl ObjectSubclass for FooPrivate {
    const NAME: &'static str = "Foo";
    type Type = Foo;
    type ParentType = glib::Object;
    type Instance = glib::subclass::simple::InstanceStruct<Self>;
    type Class = glib::subclass::simple::ClassStruct<Self>;

    glib_object_subclass!();

    fn new() -> Self {
        let x = Self {};
        let _boom = x.get_instance();
        x
    }
}
```

**Comment by sdroege** [comment]
Yes, same thing really. Fortunately this will immediately crash when you try to make use of the instance instead of causing hard to debug issues :)

**Comment by sdroege** [comment]
I think the solution here is waiting for arbitrary `Self` types to become stable. Then we could have a `Pin`-like wrapper around the references of the impl struct wherever it is used that guarantees that it's allocated correctly, and then use `fn foo(self: PinLikeThing<&Self>, ...)` for all the functions.

Without that it's going to be very unergonomic to implement.
